llama.cpp

```diff
@@ -8762,287 +8770,292 @@ static int llama_decode_internal(
-            ggml_backend_tensor_get_async(backend_res, res, logits_out.data(), (n_vocab*(n_tokens - 1))*sizeof(float), n_vocab*sizeof(float));
+            } else {
+                if (cur_token + n_tokens >= n_tokens_all) {
+                    ggml_backend_tensor_get_async(backend_res, res, logits_out, n_vocab*(n_tokens - 1)*sizeof(float), n_vocab*sizeof(float));
 #ifndef NDEBUG
-            logits_valid[0] = true;
+                    logits_valid[0] = true;
 #endif
+                }
+            }
         }
-        ggml_backend_synchronize(backend_res);
-    }
 
-    // extract embeddings
-    if (cparams.embeddings && embd) {
-        ggml_backend_t backend_embd = ggml_backend_sched_get_node_backend(lctx.sched, embd);
-        GGML_ASSERT(backend_embd != nullptr);
+        // extract embeddings
+        if (cparams.embeddings && embd) {
+            ggml_backend_t backend_embd = ggml_backend_sched_get_tensor_backend(lctx.sched, embd);
+            GGML_ASSERT(backend_embd != nullptr);
 
-        switch (cparams.pooling_type) {
-            case LLAMA_POOLING_TYPE_NONE:
-                {
-                    // extract token embeddings
-                    auto & embd_out = lctx.embd;
-
-                    if (batch.logits) {
-                        embd_out.resize(n_embd * n_tokens);
-                        for (uint32_t i = 0; i < n_tokens; i++) {
-                            if (batch.logits[i] == 0) {
-                                continue;
+            switch (cparams.pooling_type) {
+                case LLAMA_POOLING_TYPE_NONE:
+                    {
+                        // extract token embeddings
+                        auto & embd_out = lctx.embd;
+
+                        if (u_batch.logits) {
+                            //embd_out.resize(n_embd * n_tokens);
+                            for (uint32_t i = 0; i < n_tokens; i++) {
+                                if (u_batch.logits[i] == 0) {
+                                    continue;
+                                }
+                                ggml_backend_tensor_get_async(backend_embd, embd, embd_out + n_embd*(i + cur_token), (n_embd*i)*sizeof(float), n_embd*sizeof(float));
                             }
-
-                            ggml_backend_tensor_get_async(backend_embd, embd, embd_out.data() + (n_embd*i), (n_embd*i)*sizeof(float), n_embd*sizeof(float));
                         }
-                    }
-                } break;
-            case LLAMA_POOLING_TYPE_CLS:
-            case LLAMA_POOLING_TYPE_MEAN:
-                {
-                    GGML_ASSERT(strcmp(embd->name, "result_embd_pooled") == 0);
+                    } break;
+                case LLAMA_POOLING_TYPE_CLS:
+                case LLAMA_POOLING_TYPE_MEAN:
+                    {
+                        // FIXME: this may not work if the sequences are split into different batches
+                        GGML_ASSERT(n_tokens_all == n_tokens);
```

Comment:
Yes, this does not work - probably need to add comment that `n_ubatch` should be big enough to fit the prompt when using non-causal masks

Repro:

```bash
./server -m ./models/bert-bge-small/ggml-model-f16.gguf --embedding -np 8 -ngl 999 -cb --port 6900 --log-format text -b 1024 -c 4096 -v -ub 512
```

```bash
# this hits the assert because the prompt is 513 tokens

python3 ./examples/server-embd.py
```

Fix in `server`:

```diff
diff --git a/examples/server/server.cpp b/examples/server/server.cpp
index 7a210b07..d98b0c71 100644
--- a/examples/server/server.cpp
+++ b/examples/server/server.cpp
@@ -1811,7 +1811,7 @@ struct server_context {
 
                         if (slot.embedding) {
                             // this prompt is too large to process - discard it
-                            if (slot.n_prompt_tokens > n_batch) {
+                            if (slot.n_prompt_tokens > params.n_ubatch) {
                                 slot.state = SLOT_STATE_PROCESSING;
                                 slot.command = SLOT_COMMAND_NONE;
                                 slot.release();
```

Reply:
This seems to be the only case where the distinction between `n_batch` and `n_ubatch` actually matters to the applications. Support for embedding models is relatively new, so hopefully it won't cause much confusion. Do we need to do anything other than fix the check in `server`?

Reply:
I think the only thing to do in `llama.cpp` is to move the assert at the beginning of `llama_decode_internal` and apply it if `!cparama.causal_attn`. Also add message to the assert: "non-causal attention requires n_ubatch >= n_tokens"

Reply:
Does the server also check that `n_ctx >= n_prompt_tokens`? `n_batch` and `n_ubatch` may be modified by `llama_new_context_with_model` if they are larger than `n_ctx`, so checking the `gpt_params` is not enough. We may need to add a `llama_n_ubatch` function to get the actual value.